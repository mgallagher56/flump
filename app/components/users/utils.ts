import type { SupabaseClient } from '@supabase/supabase-js';
import type { TFunction } from 'i18next';

export enum SignUpActionEnum {
  LOGIN = 'logIn',
  SIGNUP = 'signUp'
}

export const getSignUpButtonText = (t: TFunction, isShowLogin: boolean, isMagicLink: boolean) => {
  if (isMagicLink) return t('sendMagicLink');
  if (isShowLogin) return t(SignUpActionEnum.LOGIN);
  return t(SignUpActionEnum.SIGNUP);
};

export const getInfoMessage = (t: TFunction, isUserRegistered: boolean, isMagicLink: boolean) => {
  if (isMagicLink) return t('magicLinkSent');
  if (isUserRegistered) return t('clickAgainToLogIn');
  return null;
}

export const getSubmitUserAuthAction = async ({
  supabaseClient,
  isShowLogin,
  isMagicLink,
  formInput
}: {
  supabaseClient: SupabaseClient;
  isShowLogin: boolean;
  isMagicLink: boolean;
  formInput: { email: string; password: string };
}) => {
  if (isMagicLink)
    return await supabaseClient.auth.signInWithOtp({
      email: formInput.email,
      options: {
        emailRedirectTo: window.location.origin + '/auth/callback'
      }
    });
  if (isShowLogin) return await supabaseClient.auth.signInWithPassword(formInput);
  return await supabaseClient.auth.signUp(formInput);
};
