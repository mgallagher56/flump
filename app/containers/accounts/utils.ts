import type { Account } from './types';

export enum AccountTypeEnum {
  CURRENT = 'Current',
  SAVING = 'Saving',
  MORTGAGE = 'Mortgage',
  LOAN = 'Loan',
  OWED = 'Owed'
}

export const isAccountTypeValid = (accountType: AccountTypeEnum, accounts: Account[]) =>
  !!accounts?.map(({ type }) => type).some((type) => type === accountType);
