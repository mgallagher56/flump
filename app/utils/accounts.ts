export const monthYearSort = (
  a: {
    month?: number;
    year?: number;
  },
  b: {
    month?: number;
    year?: number;
  }
) => {
  if (a.year === b.year) {
    return a.month - b.month;
  }
  return a.year - b.year;
};
