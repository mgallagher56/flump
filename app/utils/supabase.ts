import { createBrowserClient, createServerClient } from '@supabase/auth-helpers-remix';

import type { Database } from 'db_types';

const isServer = typeof window === 'undefined';

export const getSupaBaseUrl = (isServer: boolean): string =>
  isServer ? process.env.SUPABASE_URL : window?.env?.SUPABASE_URL;
export const getSupabaseAnonKey = (isServer: boolean): string =>
  isServer ? process.env.SUPABASE_ANON_KEY : window?.env?.SUPABASE_ANON_KEY;

const supabase = createBrowserClient<Database>(getSupaBaseUrl(isServer), getSupabaseAnonKey(isServer));

export const createSupaBaseServerClient = ({ request, response }: { request: Request; response: Response }) =>
  createServerClient<Database>(getSupaBaseUrl(isServer), getSupabaseAnonKey(isServer), { request, response });

export default supabase;
