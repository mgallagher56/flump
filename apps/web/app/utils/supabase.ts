import { createBrowserClient, createServerClient, parseCookieHeader, serializeCookieHeader } from '@supabase/ssr';

import type { Database } from 'db_types';

declare global {
  interface Window {
    env: {
      SUPABASE_URL: string;
      SUPABASE_ANON_KEY: string;
    };
  }
}

const isServer = typeof window === 'undefined';

export const getSupaBaseUrl = (isServer: boolean): string =>
  isServer ? process.env.SUPABASE_URL : window?.env?.SUPABASE_URL;
export const getSupabaseAnonKey = (isServer: boolean): string =>
  isServer ? process.env.SUPABASE_ANON_KEY : window?.env?.SUPABASE_ANON_KEY;

const supabase = createBrowserClient<Database>(getSupaBaseUrl(isServer), getSupabaseAnonKey(isServer));

export const createSupaBaseServerClient = (request: Request) => {
  const headers = new Headers();

  return createServerClient<Database>(getSupaBaseUrl(isServer), getSupabaseAnonKey(isServer), {
    cookies: {
      getAll() {
        return parseCookieHeader(request.headers.get('Cookie') ?? '');
      },
      setAll(cookiesToSet) {
        cookiesToSet.forEach(({ name, value, options }) => {
          headers.append('Set-Cookie', serializeCookieHeader(name, value, options));
        });
      }
    }
  });
};

export default supabase;
