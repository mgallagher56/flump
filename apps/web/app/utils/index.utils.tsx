import type { ReactElement } from 'react';

import type { TFunction } from 'i18next';
import FLPBox from '~/components/core/structure/FLPBox';
import type { TabData } from '~/components/core/tabs/types';
import FLPHeading from '~/components/core/typography/FLPHeading';
import type { Employee } from '~/routes/_index';

import { FallbackEnums } from './utils';

const { NA } = FallbackEnums;

export const getTabsData = (employees: Partial<Employee>[], t: TFunction): TabData[] => [
  {
    label: t('featureNum', { num: 1 }),
    value: 'feature-1',
    children: (
      <main>
        <FLPBox>
          <FLPHeading as="h1">{t('featureNum', { num: 1 })}</FLPHeading>
          {employees.map(
            (employee, index): ReactElement => (
              <FLPBox key={employee.id ?? index} style={{ display: 'flex', gap: '.5rem' }}>
                <h3>{`${employee.name ?? NA} - ${employee.department ?? NA}`}</h3>
              </FLPBox>
            )
          )}
        </FLPBox>
      </main>
    )
  },
  {
    label: t('featureNum', { num: 2 }),
    value: 'feature 2',
    children: (
      <main>
        <FLPBox>
          <FLPHeading as="h1">{t('featureNum', { num: 2 })}</FLPHeading>
        </FLPBox>
      </main>
    )
  },
  {
    label: t('featureNum', { num: 3 }),
    value: 'feature-3',
    children: (
      <main>
        <FLPBox>
          <FLPHeading as="h1">{t('featureNum', { num: 3 })}</FLPHeading>
        </FLPBox>
      </main>
    )
  }
];
