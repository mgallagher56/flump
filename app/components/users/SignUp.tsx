import type { ChangeEvent, FC, KeyboardEvent } from 'react';
import { useCallback, useEffect, useState } from 'react';

import { Form } from '@remix-run/react';
import type { User } from '@supabase/supabase-js';
import { useTranslation } from 'react-i18next';
import supabase from '~/utils/supabase';
import { AuthErrorEnums } from '~/utils/utils';

import FLPButton from '../core/buttons/FLPButton';
import FLPInput from '../core/inputs/input/FLPInput';
import FLPBox from '../core/structure/FLPBox';
import FLPText from '../core/typography/FLPText';
import { getSignUpButtonText, getSubmitUserAuthAction, SignUpActionEnum } from './utils';

const { USER_ALREADY_REGISTERED } = AuthErrorEnums;

interface SignUpProps {
  action?: SignUpActionEnum;
}

const SignUp: FC<SignUpProps> = ({ action = SignUpActionEnum.SIGNUP }) => {
  const { t } = useTranslation();

  const [userData, setUserData] = useState<User>(null);
  const [error, setError] = useState<string | null>(null);
  const [sendMagicLink, setSendMagicLink] = useState<boolean>(false);
  const [isSubmitDisabled, setIsSubmitdisabled] = useState<boolean>(false);
  const [formInput, setFormInput] = useState({ email: '', password: '' });
  const isUserRegistered = error === USER_ALREADY_REGISTERED;
  const isShowLogin = action === SignUpActionEnum.LOGIN || isUserRegistered;

  const submitButtonText = getSignUpButtonText(t, isShowLogin, sendMagicLink);

  useEffect(() => {
    setIsSubmitdisabled(sendMagicLink ? !formInput.email : !formInput.email || !formInput.password);
  }, [sendMagicLink, formInput]);

  const handleChangeFormInput = useCallback((e: ChangeEvent<HTMLInputElement>): void => {
    setFormInput((prev) => ({ ...prev, [e.target.name]: e.target.value }));
  }, []);

  const handleSubmit = useCallback(async (): Promise<void> => {
    const submitAction = getSubmitUserAuthAction({
      supabaseClient: supabase,
      formInput,
      isShowLogin,
      isMagicLink: sendMagicLink
    });
    const { data, error } = await submitAction;
    setUserData(data?.user);
    setError(error?.message);
    console.log({ data, error });
  }, [sendMagicLink, formInput, isShowLogin]);

  const handleEnter = useCallback(
    (event: KeyboardEvent<HTMLDivElement>) => {
      if (event.key === 'Enter') {
        handleSubmit();
      }
    },
    [handleSubmit]
  );

  return userData?.confirmation_sent_at ? (
    <FLPBox display="flex" flexDirection="column" gap={5}>
      <FLPText>{t('confirmEmailMsg')}</FLPText>
    </FLPBox>
  ) : (
    <Form defaultValue={''} onSubmit={(e) => e.preventDefault()}>
      <FLPBox display="flex" flexDirection="column" gap={5} onKeyUp={handleEnter}>
        <FLPInput
          label={'Email:'}
          name="email"
          type="email"
          value={formInput.email}
          error={sendMagicLink && error}
          onChange={handleChangeFormInput}
        />
        {!sendMagicLink && (
          <FLPInput
            label={'Password:'}
            name="password"
            error={error}
            type="password"
            value={formInput.password}
            onChange={handleChangeFormInput}
          />
        )}
        <FLPBox display="flex" flexDirection="column" justifyContent="center" alignItems="center" gap={4}>
          {isUserRegistered ? <FLPText>{t('clickAgainToLogIn')}</FLPText> : null}
          <FLPButton onClick={handleSubmit} type="submit" disabled={isSubmitDisabled}>
            {submitButtonText}
          </FLPButton>
          {isShowLogin && (
            <FLPButton variant="ghost" onClick={() => setSendMagicLink((prev) => !prev)}>
              {t(sendMagicLink ? 'loginWithPassword' : 'loginWithMagicLink')}
            </FLPButton>
          )}
        </FLPBox>
      </FLPBox>
    </Form>
  );
};

export default SignUp;
