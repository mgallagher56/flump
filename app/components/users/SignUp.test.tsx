import type { ReactNode } from 'react';

import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { AuthErrorEnums } from '~/utils/utils';

import SignUp from './SignUp';
import { SignUpActionEnum } from './utils';

const mocks = vi.hoisted(() => ({
  signUpSpy: vi.fn(() => ({
    data: { message: 'sign up success' },
    error: { message: AuthErrorEnums.USER_ALREADY_REGISTERED }
  })),
  signOutSpy: vi.fn(),
  signInWithPasswordSpy: vi.fn(() => ({
    data: { message: 'sign in success' },
    error: { message: 'sign in error' }
  })),
  signInWithOtp: vi.fn(() => ({
    data: {
      user: {
        confirmation_sent_at: '2021-10-10T10:10:10.000Z'
      }
    },
    error: {}
  }))
}));

vi.mock('app/utils/supabase', () => ({
  default: {
    auth: {
      signUp: mocks.signUpSpy,
      signInWithPassword: mocks.signInWithPasswordSpy,
      signOut: mocks.signOutSpy,
      signInWithOtp: mocks.signInWithOtp
    }
  }
}));

vi.mock('@remix-run/react', async () => {
  const actual: Record<string, unknown> = await vi.importActual('@remix-run/react');
  return {
    ...actual,
    Form: ({ children, onSubmit }: { children: ReactNode; onSubmit }) => <form onSubmit={onSubmit}>{children}</form>,
    useRouteData: () => ({})
  };
});

const mockEmail = 'test@test.com';
const mockPassword = 'password';

describe('<SignUp />', () => {
  it('should render as expected', () => {
    const { container } = render(<SignUp />);
    expect(container).toMatchSnapshot();
  });

  const renderAndType = async ({ action }: { action: SignUpActionEnum }) => {
    const { getByLabelText, getByText, baseElement } = render(<SignUp action={action} />);
    const emailInput = getByLabelText('Email:') as HTMLInputElement;
    const passwordInput = getByLabelText('Password:') as HTMLInputElement;
    fireEvent.change(emailInput, { target: { value: mockEmail } });
    fireEvent.change(passwordInput, { target: { value: mockPassword } });

    await waitFor(() => {
      expect(emailInput.value).toBe(mockEmail);
      expect(passwordInput.value).toBe(mockPassword);
    });

    return { baseElement, getByLabelText, getByText, passwordInput };
  };

  it('should submit the form after typing and pressing enter. Should change to login button when user is already registered', async () => {
    const { getByText, passwordInput } = await renderAndType({ action: SignUpActionEnum.SIGNUP });

    fireEvent.keyUp(passwordInput, { key: 'Enter', code: 'Enter' });

    await waitFor(() => {
      expect(mocks.signUpSpy).toHaveBeenCalled();
      expect(getByText('clickAgainToLogIn')).toBeDefined();
    });
  });

  it('should submit the form and call login function after typing and clicking the button', async () => {
    const { getByText } = await renderAndType({ action: SignUpActionEnum.LOGIN });

    const submitButton = getByText('logIn') as HTMLButtonElement;
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(mocks.signInWithPasswordSpy).toHaveBeenCalled();
    });
  });

  it('should show email confirmation message when userdata contains confirmation_sent_at param', async () => {
    const { getByText } = await renderAndType({ action: SignUpActionEnum.LOGIN });
    const magicLinkBtn = getByText('loginWithMagicLink') as HTMLButtonElement;
    const submitButton = getByText('logIn') as HTMLButtonElement;

    act(() => {
      magicLinkBtn.click();
    });

    act(() => {
      fireEvent.click(submitButton);
    });

    await waitFor(() => {
      expect(mocks.signInWithOtp).toHaveBeenCalled();
      expect(getByText('confirmEmailMsg')).toBeDefined();
    });
  });
});
