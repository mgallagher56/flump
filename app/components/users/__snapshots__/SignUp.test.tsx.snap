// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SignUp /> > should render as expected 1`] = `
<div>
  <form>
    <div
      class="css-1gnc3qn"
    >
      <div
        class="d_flex flex_column gap_2"
        data-part="root"
        data-scope="editable"
        id="editable::r0:"
      >
        <label
          data-part="label"
          data-scope="editable"
          for="editable::r0::input"
          id="editable::r0::label"
        >
          Email:
        </label>
        <div
          data-part="area"
          data-placeholder-shown=""
          data-scope="editable"
          id="editable::r0::area"
        >
          <div
            class="css-3a3hw0"
          >
            <input
              aria-label="editable input"
              class="chakra-input css-0"
              data-part="input"
              data-scope="editable"
              id="editable::r0::input"
              name="email"
              type="email"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="d_flex flex_column gap_2"
        data-part="root"
        data-scope="editable"
        id="editable::r1:"
      >
        <label
          data-part="label"
          data-scope="editable"
          for="editable::r1::input"
          id="editable::r1::label"
        >
          Password:
        </label>
        <div
          data-part="area"
          data-placeholder-shown=""
          data-scope="editable"
          id="editable::r1::area"
        >
          <div
            class="css-3a3hw0"
          >
            <input
              aria-label="editable input"
              class="chakra-input css-0"
              data-part="input"
              data-scope="editable"
              id="editable::r1::input"
              name="password"
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="css-w98boq"
      >
        <button
          class="chakra-button css-4xx2wk"
          id="pressable::r2:"
          type="submit"
        >
          signUp
        </button>
      </div>
    </div>
  </form>
</div>
`;
